xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -70,9 +83,109 @@ ConfigUpdate build() {
    */
-  // TODO(zdapeng): content TBD.
   static final class ClusterUpdate {
+    private final String clusterName;
+    private final String edsServiceName;
+    private final String lbPolicy;
+    private final boolean enableLrs;
+    private final String lrsServerName;
+
+    private ClusterUpdate(String clusterName, String edsServiceName, String lbPolicy,
+        boolean enableLrs, @Nullable String lrsServerName) {
+      this.clusterName = clusterName;
+      this.edsServiceName = edsServiceName;
+      this.lbPolicy = lbPolicy;
+      this.enableLrs = enableLrs;
+      this.lrsServerName = lrsServerName;
+    }
+
+    String getClusterName() {
+      return clusterName;
+    }
+
+    /**
+     * Returns the resource name for EDS requests.
+     */
+    String getEdsServiceName() {
+      return edsServiceName;
+    }
+
+    /**
+     * Returns the policy of balancing loads to endpoints. Always returns "round_robin".
+     */
+    String getLbPolicy() {
+      return lbPolicy;
+    }
+
+    /**
+     * Returns true if LRS is enabled.
+     */
+    boolean isEnableLrs() {
+      return enableLrs;
+    }
+
+    /**
+     * Returns the server name to send client load reports to if LRS is enabled. {@code null} if
+     * {@link #isEnableLrs()} returns {@code false}.
+     */
+    @Nullable
+    String getLrsServerName() {
+      return lrsServerName;
+    }
+
+    static Builder newBuilder() {
+      return new Builder();
+    }
+
+    static final class Builder {
+      private String clusterName;
+      private String edsServiceName;
+      private String lbPolicy;
+      private boolean enableLrs;
+      @Nullable
+      private String lrsServerName;
+
+      // Use ClusterUpdate.newBuilder().
+      private Builder() {
+      }
+
+      Builder setClusterName(String clusterName) {
+        this.clusterName = clusterName;
+        return this;
+      }
+
+      Builder setEdsServiceName(String edsServiceName) {
+        this.edsServiceName = edsServiceName;
+        return this;
+      }
+
+      Builder setLbPolicy(String lbPolicy) {
+        this.lbPolicy = lbPolicy;
+        return this;
+      }
+
+      Builder setEnableLrs(boolean enableLrs) {
+        this.enableLrs = enableLrs;
+        return this;
+      }
 
+      Builder setLrsServerName(String lrsServerName) {
+        this.lrsServerName = lrsServerName;
+        return this;
+      }
+
+      ClusterUpdate build() {
+        Preconditions.checkState(clusterName != null, "clusterName is not set");
+        Preconditions.checkState(lbPolicy != null, "lbPolicy is not set");
+        Preconditions.checkState(
+            (enableLrs && lrsServerName != null) || (!enableLrs && lrsServerName == null),
+            "lrsServerName is not set while LRS is enabled "
+                + "OR lrsServerName is set while LRS is not enabled");
+        return
+            new ClusterUpdate(clusterName, edsServiceName == null ? clusterName : edsServiceName,
```

Comment:
This logic need unit-test coverage.

Reply:
Ok. Isn't this review being too strict? :)

Reply:
I tried to be not too strict by not asking for test  `(enableLrs && lrsServerName != null) || (!enableLrs && lrsServerName == null)` :)

Reply:
We had the check. If the invariant is violated, it will break immediately and we will know.

Reply:
That's if the invariant is violated. But `checkState(complicatedFunction(arg1, arg2))` will never break immediately and let you know if `complicatedFunction(arg1, arg2)` has a bug and always returns `true`.

Reply:
This is the invariant we put for users using this class, it is the restriction on how users should use it. The class implementation is not wrong, it just sets whatever passed in. Even if we added test for validating this invariant-check works, it still does not guarantee users will not break it. So this is the place we put invariant-check instead of unit test.